src/python/pants/engine/rules.py

```diff
@@ -506,10 +523,14 @@ def dependency_optionables(self):
     return tuple()
 
 
-# TODO: add typechecking here -- would need to have a TypedCollection for dicts for `union_rules`.
-class RuleIndex(datatype(['rules', 'roots', 'union_rules'])):
+@dataclass(frozen=True)
+class RuleIndex:
   """Holds a normalized index of Rules used to instantiate Nodes."""
 
+  rules: Dict[Any, Any]
```

Comment:
Can this be tightened at all? I think rules are seen as `Callable`? Don't let this block landing the PR though. We can come back to this and I'd rather land the great work you've done.

Reply:
This is a better version of lines 573-575 - should delete the original.

Thanks for adding type hints! Is `Any` representative or because we don't know how to make it more precise? If the latter, it's better to do `Dict` and `Set` with no types specified so that we know to later go back and make the type hint more precise.

Reply:
I have a followup PR that addresses those more-precise type hints!

Reply:
See #8496!